### PR TITLE
Launch dockerd on lima

### PR DESCRIPTION
**Pull request by ericpromislow**
This PR does launch `dockerd` on lima, and modifies the user config to forward ports, but we get permission-denied errors when trying to access the dockerd socket from the client, even after adding the VM user to the `docker` group.

Creating the PR now in case the permission work should be done in a later task. Note that this PR _does_ launch `dockerd` on lima, as per the label.

### CODE REVIEW

File: src/k8s-engine/lima.ts [hunk 1]
```diff
@@ -453,6 +455,27 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
...
+      config.portForwards = allPortForwards = DEFAULT_CONFIG.portForwards ?? [];
+    }
+    const dockerPortForwards = allPortForwards?.find(pair => Object.keys(pair).length === 2 &&
+      pair.guestSocket === '/run/docker.sock' &&
+      ('hostSocket' in pair));
+
+    if (!dockerPortForwards) {
+      config.portForwards?.push({
+        guestSocket: '/run/docker.sock',
+        hostSocket: path.join(paths.lima, MACHINE_NAME, 'docker.sock')
```
Comment (jandubois) [inline src/k8s-engine/lima.ts:472]: Please make sure that any symlinks in `path.lima` have been evaluated. If the host username is too long, the lima directory may need to be replaced with a symlink to something with a shorter path.

File: src/k8s-engine/lima.ts [hunk 2]
```diff
@@ -866,6 +889,19 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
           return;
         }
 
+        const username = process.env.USER || process.env.LOGNAME || process.env.USERNAME;
```
Comment (jandubois) [inline src/k8s-engine/lima.ts:892]: This sets the host username. You need the username inside the guest. That may be different when the macOS username is not a valid Linux username. In that case the guest username will be `lima`.

You should be able to get the name in the guest via `${USER}` or `$(id -nu)`.

**Comment by jandubois** [comment]
> we get permission-denied errors when trying to access the dockerd socket from the client, even after adding the VM user to the `docker` group.

Ok, it's complicated... 😞 

In lima we have a persistent user session, so the group set is never getting updated, even if we log out and back in. You can create a new session with `newgrp docker`, and you will have access to the socket in that session, but there is no way to update the session set for the ssh process maintaining the control channel.

I thought rebooting would help (even though inconvenient), but it doesn't because we are using an Alpine ISO, so recreate the user during cloud-init. We later switch out `/etc` to the persisted copy on `/mnt/data/etc`, but this is again too late to set the group set of the ssh process.

I think the best way forward right now is simply to make the socket inside the VM world-writable:

```
sudo chmod 666 /var/run/docker.sock
```

Effectively this is the same as adding `docker` to the group set, as we don't really have any less-privileged users that we want to prevent from accessing the socket.

The forwarded socket on the host will still have `600` permission anyways.

**Review by mook-as — CHANGES_REQUESTED** [review]
Do we not need relevant knobs? This currently starts `dockerd` _and_ `containerd` (via `k3s`), which is probably not what we want.

File: src/k8s-engine/lima.ts (hunk 2 above)
Comment (ericpromislow) [reply src/k8s-engine/lima.ts:892]: This isn't needed if we make the socket world-accessible

File: src/k8s-engine/lima.ts (hunk 1 above)
Comment (ericpromislow) [reply src/k8s-engine/lima.ts:472]: Why is it an issue if `paths.lima` is a symlink? This is on the host side, what's the problem?

Comment (jandubois) [reply src/k8s-engine/lima.ts:472]: Explained on slack, but in short: the socket path must be less than 104 characters on macOS (108 on Linux)

File: src/k8s-engine/lima.ts [hunk 3]
```diff
@@ -453,6 +455,27 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
...
   }
 
+  protected updateConfigPortForwards(config: LimaConfiguration) {
+    let allPortForwards: Array<Record<string, any>> | undefined = config.portForwards;
+
+    if (!allPortForwards) {
+      // This shouldn't happen, but fix it anyway
+      config.portForwards = allPortForwards = DEFAULT_CONFIG.portForwards ?? [];
+    }
+    const dockerPortForwards = allPortForwards?.find(pair => Object.keys(pair).length === 2 &&
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/k8s-engine/lima.ts]: If someone explicity changes the config to put `hostSocket` at a different location, I would respect that and leave it. If the difference is due to corruption of the config file, we're wading into an area where software providers often will recommend shutting down, deleting the config file, and restarting.

**Comment by ericpromislow** [comment]
We still need the UI for choosing between dockerd and containerd, that's a separate issue. Given that I was able to pull and access images using docker with this change, and can still use images in the UI, they aren't interfering with one another. But sure, we do want to run only one at a time.

File: src/k8s-engine/lima.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jandubois) [reply src/k8s-engine/lima.ts]: Users should not need to modify the config file. They should move the whole `lima` directory to a different location, if they have to. This could even be to a separate disk, in case there is not enough free space on the host boot disk for the VM (grows to 100GB).

File: src/k8s-engine/lima.ts [hunk 4]
```diff
@@ -926,6 +984,11 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
           100,
           this.client?.waitForReadyNodes() ?? Promise.reject(new Error('No client')));
 
+        // The socket isn't present immediately after `docker start` finishes, but it should be by this point
```
Comment (jandubois) [inline src/k8s-engine/lima.ts:987]: This comment sets off alarm bells for me; it basically says: "we have a race condition here, but I think we will be fine".

I think the proper way should be to wait until the socket is available, and *then* change permissions.

I wonder if this could be done inside the provisioning script:

```shell
rc-service docker start
while ! [ -S /var/run/docker.sock ]; do sleep 1; done
chmod 666 /var/run/docker.sock
```

File: src/k8s-engine/lima.ts [hunk 5]
```diff
@@ -467,9 +495,12 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
       ('hostSocket' in entry));
 
     if (!dockerPortForwards) {
+      const hostSocketPath = `${ paths.lima }/${ MACHINE_NAME }/docker.sock`;
+
+      await this.checkExpandedLength(hostSocketPath);
       config.portForwards?.push({
         guestSocket: '/run/docker.sock',
-        hostSocket:  `{{.LimaHome}}/${ MACHINE_NAME }/docker.sock`,
+        hostSocket:  hostSocketPath,
```
Comment (jandubois) [inline src/k8s-engine/lima.ts:503]: I think this is still the original path, and not the one with the expanded symlinks, right? You need to forward to the valid socket path.

File: src/k8s-engine/lima.ts [hunk 6]
```diff
@@ -455,7 +455,35 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
     }
   }
 
-  protected updateConfigPortForwards(config: LimaConfiguration) {
+  protected async checkExpandedLength(proposedPath: string) {
```
Comment (jandubois) [inline src/k8s-engine/lima.ts:458]: I think `checkExpandedLength` is a bit obscure. You really need to read the function logic to understand it is about the length of the path after symlinks have been expanded.

I would split this into 2 functions: `EvaluateSymlinks` and `CheckMaxSocketLength` or similar, as these are really unrelated functionalities.

File: src/k8s-engine/lima.ts (hunk 5 above)
Comment (ericpromislow) [reply src/k8s-engine/lima.ts:503]: Ah, I thought we just had to verify the length of the symlink-evaluated path was within limits.

File: src/k8s-engine/lima.ts (hunk 4 above)
Comment (ericpromislow) [reply src/k8s-engine/lima.ts:992]: Yes, that sounds better.  I ssh'ed in and ran a `watch ls /run` command and it was created shortly after the server was started, well before the K8s API was ready, but scripting it would be better.

**Comment by jandubois** [comment]
I've tested the symlink support and it works:

```console
$ cd ~/Library/Application\ Support/rancher-desktop/
$ mv lima ~/.rdlima
$ ln -s ~/.rdlima lima
$ docker -H unix://$HOME/.rdlima/0/docker.sock images
$ docker -H unix://$HOME/Library/Application\ Support/rancher-desktop/lima/0/docker.sock images
```

Somewhat to my surprise the docker cli works fine when you pass it the path with symlinks not yet evaluated. Which makes me think that maybe lima should do the `EvalSymlinks` for the hostpath too.

But for now what we have should be good!

**Review by jandubois — APPROVED** [review]
LGTM now, but please check with @mook-as if the "knobs" to select between `containerd` and `dockerd` should be part of this PR.

I think it would be part of #814

**Review by mook-as — CHANGES_REQUESTED** [review]
Just stuff dealing with symlinks / proposed path.

File: src/k8s-engine/lima.ts [hunk 7]
```diff
@@ -453,6 +455,63 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
...
+    return actualPath;
+  }
+
+  protected checkMaxSocketLength(proposedPath: string) {
+    // See https://serverfault.com/questions/641347/check-if-a-path-exceeds-maximum-for-unix-domain-socket
+    // for an example of how to determine these values.
+    const socketLengthLimit = os.platform() === 'darwin' ? 103 : 107;
+
+    if (proposedPath.length > socketLengthLimit) {
+      console.log(`Specified path ${ proposedPath } symlink-expands to ${ proposedPath }`);
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/k8s-engine/lima.ts:485]: typo, good catch

File: src/k8s-engine/lima.ts [hunk 8]
```diff
@@ -453,6 +455,63 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
...
 
+  protected async evalSymlinks(proposedPath: string) {
+    const dirs = proposedPath.split(path.sep);
+    let actualPath = '/';
+
+    for (let i = 1; i < dirs.length; i++) {
+      const currentPath = path.join(actualPath, dirs[i]);
+
+      try {
+        actualPath = path.resolve(actualPath, await fs.promises.readlink(currentPath));
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/k8s-engine/lima.ts]: This variant was more elegant, but as per the slack conversation that Lima needs to do this anyway, we're just going to set the `hostSocket` field to `docker` and let Lima do all the checking and reporting.

File: src/k8s-engine/lima.ts [hunk 9]
```diff
@@ -453,6 +455,63 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
...
+        actualPath = currentPath;
+      }
+    }
+
+    return actualPath;
+  }
+
+  protected checkMaxSocketLength(proposedPath: string) {
+    // See https://serverfault.com/questions/641347/check-if-a-path-exceeds-maximum-for-unix-domain-socket
+    // for an example of how to determine these values.
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/k8s-engine/lima.ts]: No longer needed.

**Review by ericpromislow — COMMENTED** [review]
Not sure how github turned my comments into a review, but it's not.

**Review by mook-as — APPROVED** [review]
